[Structural] change of convergence criteria and test

RiccardoRossi opened the pull request:
Currently the patch_test large strain was always iterating to the max number of iterations. this is so since external forces are zero and the residual is approx zero in the first iteration, so convergence was never detected (even though the result was correct).
Due to this reason i am proposing changing ResidualCriteria->DisplacementCriteria

as an aside i am taking the occasion to use the BlockBuilderAndSolver instead of the Elimination version, since this is used by default by the structural solver

RiccardoRossi committed:
change of convergence criteria and test

Currently the patch_test large strain was always iterating to the max number of iterations. this is so since external forces are zero and the residual is approx zero in the first iteration, so convergence was never detected (even though the result was correct).
Due to this reason i am proposing changing ResidualCriteria->DisplacementCriteria

as an aside i am taking the occasion to use the BlockBuilderAndSolver instead of the Elimination version, since this is used by default by the structural solver

philbucher commented on applications/StructuralMechanicsApplication/tests/test_patch_test_large_strain.py:
```diff
@@ -104,9 +104,9 @@ def _solve(self,mp):
         #define a minimal newton raphson solver
         linear_solver = KratosMultiphysics.SkylineLUFactorizationSolver()
-        builder_and_solver = KratosMultiphysics.ResidualBasedEliminationBuilderAndSolver(linear_solver)
+        builder_and_solver = KratosMultiphysics.ResidualBasedBlockBuilderAndSolver(linear_solver)
```
do you really want to change this?
I would leave it, since otherwise we have almost no tests using it
The BlockBuilder is used everywhere

RiccardoRossi replied:
well, maybe i can modify the test to use one or the other. point is that i ultimately want to use this as a test also for the future PR on using the old stiffness at first iteration

philbucher replied:
+1 for using both

philbucher commented:
@RiccardoRossi since you are at it, could you add a check for the number of NL-iterations such that this does not happen again
i.e. sth like:
~~~py
self.assertEqual(5, mp.ProcessInfo[KM.NL_ITERATIONS])
~~~

RiccardoRossi commented:
> @RiccardoRossi since you are at it, could you add a check for the number of NL-iterations such that this does not happen again
> i.e. sth like:
> 
> ```python
> self.assertEqual(5, mp.ProcessInfo[KM.NL_ITERATIONS])
> ```

I am not sure about this check. the point is that the exact number of iterations will depend on the solution options so i am not completely sure if this should be done or not. 

In any case i have some troubles right now since one of the UL tests fails for (yet) unknown reasons when using the UL element.

RiccardoRossi committed:
splitting

ok, this one calls the function twice resetting in between.

the problem si that the UL appears to fail if the solve is called more than once, which is definitely a BIG problem

@loumalouomega we need you to take a look at this, one of the requirements is to be able to call the solve multiple times with the same results (as long as the material is elastic). 
As a rule, this should also be the case for plastic materials IF THE FINALIZE SOLUTION STEP FUNCTION is not called (probably not relevant here)

RiccardoRossi commented on applications/StructuralMechanicsApplication/tests/test_patch_test_large_strain.py:
```diff
@@ -376,15 +400,15 @@ def test_compare_TL_UL_2D_triangle(self):
-                    self.assertLess((tl_dx - ul_dx) / ul_dx, 1.0e-3)
-                if (ul_dy > 0.0):
-                    self.assertLess((tl_dy - ul_dy) / ul_dy, 1.0e-3)
+                if(abs(tl_dx) > 1e-20):
```
the relative difference between TL and UL should be zero for an elastic material up to machine precision...

@loumalouomega

loumalouomega replied:
Ok

RiccardoRossi commented on applications/StructuralMechanicsApplication/tests/test_patch_test_large_strain.py:
```diff
@@ -419,8 +443,15 @@ def test_TL_2D_triangle(self):
+        self._ResetDisplacementAndPosition(mp)
+        self._apply_BCs(bcs,A,b)
+        self._solve(mp,"block_builder")
+        self._check_results(mp,A,b)
```
note that this check fails even if you simply call the _solve function twice, independently on which builder and solver is used.

RiccardoRossi replied:
i would say it still fails. just try to copy and paste the line 426 just below

philbucher commented:
> I am not sure about this check. the point is that the exact number of iterations will depend on the solution options so i am not completely sure if this should be done or not.

well this is a test where you control those variables, so I think it makes sense
Or at least check that the number of nl-iterations don't exceed a certain number

loumalouomega commented on applications/StructuralMechanicsApplication/tests/test_patch_test_large_strain.py:
```diff
@@ -340,7 +359,10 @@ def test_compare_TL_UL_2D_triangle(self):
         delta_time = ul_mp.ProcessInfo[KratosMultiphysics.DELTA_TIME]
         time = ul_mp.ProcessInfo[KratosMultiphysics.TIME]
 
+        print("360")
```
Remove this

loumalouomega commented on applications/StructuralMechanicsApplication/tests/test_patch_test_large_strain.py:
```diff
@@ -340,7 +359,10 @@ def test_compare_TL_UL_2D_triangle(self):
+        print("360")
         tl_strategy = self._create_strategy(tl_mp)
+
+        print("363")
```
And this

loumalouomega commented on applications/StructuralMechanicsApplication/tests/test_patch_test_large_strain.py:
```diff
@@ -367,7 +389,9 @@ def test_compare_TL_UL_2D_triangle(self):
                 #node.SetSolutionStepValue(KratosMultiphysics.DISPLACEMENT_Y, iter * 5.0e-1)
 
             self._solve_with_strategy(tl_strategy, tl_lhs, iter)
+            print("a")
```
so many prints

RiccardoRossi commented:
ok i agree, will do (i mean, the check on iterations)

RiccardoRossi commented:
@loumalouomega the point is that this is currently failing, and the prints were needed to see where it stops.

as i am telling the failure comes if you simply run twice the _solve.

loumalouomega committed:
Merge branch 'master' into Changing-convergence-criteria-and-builderandsolver-in-test

loumalouomega committed:
Removing prints

loumalouomega committed:
Reduce code duplication

loumalouomega committed:
Fix test

loumalouomega commented:
@RiccardoRossi the problem was not the UL, the problem is that python does not copy, it uses references, therefore A and b change

RiccardoRossi committed:
using F=A.copy()

it makes no sense that A is changed in the check...this should solve the problem

RiccardoRossi reviewed (commented):
you are right about the copy, however i think there is stil la problem. But i still think there are some underling issues

loumalouomega commented:
The thing is that the TL does not break because it integrates in the
reference configuration, which is not broken

El dom., 22 mar. 2020 16:39, Riccardo Rossi <notifications@github.com>
escribió:

> *@RiccardoRossi* commented on this pull request.
>
> you are right about the copy, however i think there is stil la problem.
> But i still think there are some underling issues
> ------------------------------
>
> In
> applications/StructuralMechanicsApplication/tests/test_patch_test_large_strain.py
> <https://github.com/KratosMultiphysics/Kratos/pull/6572#discussion_r395988688>
> :
>
> > @@ -419,8 +443,15 @@ def test_TL_2D_triangle(self):
>
>          A,b = self._define_movement(dim)
>
> +        self._ResetDisplacementAndPosition(mp)
> +        self._apply_BCs(bcs,A,b)
> +        self._solve(mp,"block_builder")
> +        self._check_results(mp,A,b)
>
> i would say it still fails. just try to copy and paste the line 426 just
> below
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/KratosMultiphysics/Kratos/pull/6572#pullrequestreview-378915191>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEYQZAGRT3XOFMO4NGJ3T4LRIYWLRANCNFSM4LPDOF7A>
> .
>

RiccardoRossi commented:
@loumalouomega the point of this PR is exactly to push the database back to the beginning of the step in computing the LHS and RHS (at the first iteration). this implies that within SolveSolutionStep the integration IS done on the reference configuration

BUT

the problem is that right now the test does:

- new step
- ApplyBC --> here we move the mesh
- InitializesolutionStep --> so initialization, and hence the reference, is on the deformed config
- SolveSolutionStep  --> so if here i push back, i am pushing back to BEFORE ApplyBC, so to a configuraiton that is different from the one on which InitializeSolutionStep was done

the proposal of #6567 is to change this to 

- new step
- InitializesolutionStep --> so initialization, and hence the reference, is on the UNDEFORMED config
- ApplyBC --> here we move the mesh
- SolveSolutionStep  --> so if here i push back, i am pushing back to BEFORE ApplyBC, so correctly to the reference configuration

loumalouomega commented:
> @loumalouomega the point of this PR is exactly to push the database back to the beginning of the step in computing the LHS and RHS (at the first iteration). this implies that within SolveSolutionStep the integration IS done on the reference configuration
> 
> BUT
> 
> the problem is that right now the test does:
> 
>     * new step
> 
>     * ApplyBC --> here we move the mesh
> 
>     * InitializesolutionStep --> so initialization, and hence the reference, is on the deformed config
> 
>     * SolveSolutionStep  --> so if here i push back, i am pushing back to BEFORE ApplyBC, so to a configuraiton that is different from the one on which InitializeSolutionStep was done
> 
> 
> the proposal of #6567 is to change this to
> 
>     * new step
> 
>     * InitializesolutionStep --> so initialization, and hence the reference, is on the UNDEFORMED config
> 
>     * ApplyBC --> here we move the mesh
> 
>     * SolveSolutionStep  --> so if here i push back, i am pushing back to BEFORE ApplyBC, so correctly to the reference configuration

@RiccardoRossi test is failing because copy of the matrix is not implemented

loumalouomega committed:
Merge branch 'master' into Changing-convergence-criteria-and-builderandsolver-in-test

loumalouomega committed:
Clean up
